### PR TITLE
8295867: TestVerifyGraphEdges.java fails with exit code -1073741571 when using AlwaysIncrementalInline

### PR DESCRIPTION
**Pull request by vnkozlov**
When we use -Xcomp we compile `java.lang.invoke.LambdaForm$Kind::<clinit>` very long linear method for enum class: [LambdaForm.java#L250](https://github.com/openjdk/jdk/blame/master/src/java.base/share/classes/java/lang/invoke/LambdaForm.java#L250)

In addition we inline all <init> class initializers for EA when run with -Xcomp: [bytecodeInfo.cpp#L410](https://github.com/openjdk/jdk/blame/master/src/hotspot/share/opto/bytecodeInfo.cpp#L410)

Recent CDS change [JDK-8293979](https://bugs.openjdk.org/browse/JDK-8293979) allows to inline a bit more deeply too.

Adding `-XX:+AlwaysIncrementalInline` worsened situation even more.

Running with `-XX:+LogCompilation` shows that we hit `NodeCountInliningCutoff (18000)` during `java.lang.invoke.LambdaForm$Kind::<clinit>` compilation.

In short, we have very long (>40000 live nodes) linear IR graph. `Node::verify_edges()` method process nodes depth-first starting from first input which is control edge. So it is not surprise that depth of this method recursion reached 6000.
With frame size of 10 words (320 bytes) we easy hit stack overflow (768K in 32-bits debug VM).

I fixed it by using local buffer `Node_List` instead of recursion in `Node::verify_edges()`.
The algorithm was changed to simplify code. It processes inputs in reverse order - last input processed first. And I noticed that maximum use of buffer is only about 1000 or less elements for this compilation (that is why I use live_nodes/16 as initial size of buffer).

Then I did additional experiment with keeping recursion but processing inputs in reverse order:

```
   // Recursive walk over all input edges
-  for( i = 0; i < len(); i++ ) {
-    n = in(i);
+  for( i = len(); i > 0; i++ ) {
+    n = in(i - 1);
     if( n != NULL )
       in(i)->verify_edges(visited);
   }
```

And it shows the same around 1000 stack depth!

I decided to keep my original fix because it should be faster (put only one value on list instead of putting all locals, PC, SP on stack and calls) and much less stack usage.

Testing tier1-3, hs-comp-stress and `TestVerifyGraphEdges.java` test runs with `-XX:+AlwaysIncrementalInline`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295867](https://bugs.openjdk.org/browse/JDK-8295867): TestVerifyGraphEdges.java fails with exit code -1073741571 when using AlwaysIncrementalInline

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11065/head:pull/11065` \
`$ git checkout pull/11065`

Update a local copy of the PR: \
`$ git checkout pull/11065` \
`$ git pull https://git.openjdk.org/jdk pull/11065/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11065`

View PR using the GUI difftool: \
`$ git pr show -t 11065`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11065.diff">https://git.openjdk.org/jdk/pull/11065.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/node.cpp [hunk 1]
```diff
@@ -2683,46 +2683,52 @@ void Node::dump_comp(const char* suffix, outputStream *st) const {
...
+        int cnt = 0;
+        for (uint idx = 0; idx < n->_outcnt; idx++) {
+          if (n->_out[idx] == next) {
+            cnt++;
+          }
+        }
+        assert(cnt > 0, "Failed to find Def-Use edge.");
+        // Check for duplicate edges
+        // walk the input array downcounting the input edges to n
+        for(uint j = 0; j < length; j++) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/node.cpp:2714]:
```suggestion
        for (uint j = 0; j < length; j++) {
```

File: src/hotspot/share/opto/node.cpp [hunk 2]
```diff
@@ -2683,46 +2683,52 @@ void Node::dump_comp(const char* suffix, outputStream *st) const {
...
+    Node* next = nstack.pop();
+    if (visited.member(next)) {
+      continue;
+    }
+    visited.push(next);
+
+    // Walk over all input edges, checking for correspondence
+    uint length = next->len();
+    for (uint i = 0; i < length; i++) {
+      Node* n = next->in(i);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/node.cpp:2699]: I suggest to rename `n` to `input` to make it easier to see if it is the current node or an input to it.

File: src/hotspot/share/opto/node.hpp [hunk 3]
```diff
@@ -1217,7 +1217,7 @@ class Node {
   // Print compact per-node info
   virtual void dump_compact_spec(outputStream *st) const { dump_spec(st); }
 
-  void verify_edges(Unique_Node_List &visited); // Verify bi-directional edges
+  static void verify_edges(Node* root, Unique_Node_List &visited, Node_List &nstack); // Verify bi-directional edges
```
Comment (chhagedorn) [inline src/hotspot/share/opto/node.hpp:1220]: Maybe you can directly rename the method according to your comment: `verify_bidirectional_edges()`.

**Review by chhagedorn — APPROVED** [review]
I agree that a non-recursive solution is preferable in this case. I only have some minor code style comments - otherwise, the fix looks good!

**Review by shipilev — APPROVED** [review]
x86_32 seems to be happy with this change. I ran both `TestVerifyGraphEdges` and `tier1 tier2` with `-XX:+VerifyGraphEdges` without problems with Linux x86_32 fastdebug. The code looks reasonable too.

File: src/hotspot/share/opto/node.cpp (hunk 2 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/node.cpp:2699]: Renamed to `in`

File: src/hotspot/share/opto/compile.cpp [hunk 4]
```diff
@@ -4238,8 +4238,12 @@ bool Compile::needs_clinit_barrier(ciInstanceKlass* holder, ciMethod* accessing_
 void Compile::verify_graph_edges(bool no_dead_code) {
   if (VerifyGraphEdges) {
     Unique_Node_List visited;
+    // Allocate stack of size C->live_nodes()/16 to avoid frequent realloc
+    uint stack_size = live_nodes() >> 4;
+    Node_List nstack(MAX2(stack_size, (uint)OptoNodeListSize));
```
[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/share/opto/compile.cpp:4243]: I need `live_nodes()` value or `stack_size` or `C` to pass for creating list inside method.
I decided to move renamed `verify_bidirectional_edges()` method to `Compile` class to get this values inside the method.
It does not need to be in `Node` class after I removed recursion.

**Review by shipilev — APPROVED** [review]

File: src/hotspot/share/opto/compile.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (chhagedorn) [reply src/hotspot/share/opto/compile.cpp]: Right, I've missed that before. Moving it to `Compile` is a good idea to apply that!

**Review by chhagedorn — APPROVED** [review]
Thanks for doing the updates, looks good!

**Pull request closed without merging**